Improve UWP view presenter to use presentation attributes v2

nickrandolph opened the pull request:
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Changing the UWP view presenter to make use of new attribute system.

### :arrow_heading_down: What is the current behavior?
Presenter is not able to make us of attributes.

### :new: What is the new behavior (if this is a feature change)?


### :boom: Does this PR introduce a breaking change?
Yes, UWP Pages need annotations now 

### :bug: Recommendations for testing


### :memo: Links to relevant issues/docs
This supersedes the existing PR: https://github.com/MvvmCross/MvvmCross/pull/2348

### :thinking: Checklist before submitting

- [x] All projects build
- [x] Follows style guide lines ([code style guide](https://github.com/MvvmCross/MvvmCross#code-style-guidelines))
- [ ] Relevant documentation was updated ([docs style guide](https://www.mvvmcross.com/documentation/contribute/mvvmcross-docs-style-guide))
- [x] Nuspec files were updated (when applicable)
- [x] Rebased onto current develop

Known issues: 
XForms for UWP doesn't work and is not fixed in the PR
There is still some clean up to be done around multiple presenters
Playground.UWP still references XForms - removing this causes an exception. Need to separate UI/ViewModels into separate projects and fix this

Cheesebaron commented:
@nickrandolph can you take a look at the issues CodeFactor is pointing out?

nickrandolph commented:
@Cheesebaron  I've addressed the obvious layout issues raised by CodeFactor. The remaining issues are due to the duplication that was already present with the viewpresenters. This PR already reduces this with the MvxAttributeViewPresenter. I'm not familiar enough with the other platforms to know how to refactor the code from those platforms where there is still code duplicity

martijn00 commented on MvvmCross/Core/Core/Views/MvxAttributeViewPresenter.cs:
```diff
@@ -0,0 +1,97 @@
+
+        public abstract MvxBasePresentationAttribute CreatePresentationAttribute(Type viewModelType, Type viewType);
+
+        public virtual MvxBasePresentationAttribute GetOverridePresentationAttribute(Type viewModelType, Type viewType)
```
Should this not be abstract as well?

nickrandolph replied:
@martijn00 it looked like this was a common implementation for this method, so think prob reasonable to have as base implementation. Can make abstract but then we'll end up with code duplication across the platform implementations

martijn00 commented on MvvmCross/Core/Core/Views/MvxAttributeViewPresenter.cs:
```diff
@@ -0,0 +1,97 @@
+
+namespace MvvmCross.Core.Views
+{
+    public abstract class MvxAttributeViewPresenter : MvxViewPresenter, IMvxAttributeViewPresenter
```
Good idea to do this!

martijn00 commented on MvvmCross/Core/Core/Platform/LogProviders/ConsoleLogProvider.cs:
```diff
@@ -178,7 +178,8 @@ private static class ConsoleColorValues
             static ConsoleColorValues()
             {
-                Type = Type.GetType("System.ConsoleColor");
+                Type = Type.GetType("System.ConsoleColor") ?? Type.GetType("System.ConsoleColor, System.Console");
```
@willsb what do you think of this?

nickrandolph replied:
I was actually thinking we can drop the initial GetType and just use the assembly qualified call, but will that work across all platforms?

Grrbrr404 replied:
When I started working on this I had issues with  Type.GetType("System.ConsoleColor") on runtime, always returning null. So the workarround with Type.GetType("System.ConsoleColor, System.Console"); worked for me.

martijn00 commented on MvvmCross/Windows/Uwp/Views/MvxWindowsViewPresenter.cs:
```diff
@@ -28,34 +33,111 @@ public MvxWindowsViewPresenter(IMvxWindowsFrame rootFrame)
+                typeof(MvxPagePresentationAttribute),
+                new MvxPresentationAttributeAction
+                {
+                    ShowAction = (view, attribute, request) => ShowPage(request),
```
It's probably best to send the view, attribute and request parameters already to this method. You might not need them now, but maybe later, or someone overriding this needs it.

martijn00 commented on MvvmCross/Windows/Uwp/Views/MvxWindowsViewPresenter.cs:
```diff
@@ -28,34 +33,111 @@ public MvxWindowsViewPresenter(IMvxWindowsFrame rootFrame)
+        public override MvxBasePresentationAttribute CreatePresentationAttribute(Type viewModelType, Type viewType)
+        {
+            MvxTrace.Trace($"PresentationAttribute not found for {viewType.Name}. Assuming new page presentation");
+            return new MvxPagePresentationAttribute();
```
Is there no way to detect a splitview?

nickrandolph replied:
@martijn00 not without creating an instance of the viewType - not sure we should be encouraging that behaviour from developers as it'll only be negative for perf

martijn00 commented on MvvmCross/Windows/Uwp/Views/MvxWindowsViewPresenter.cs:
```diff
@@ -109,49 +173,86 @@ protected virtual string GetRequestText(MvxViewModelRequest request)
         }
 
-        protected virtual void HandleBackButtonVisibility()
+        private void ShowPage(MvxViewModelRequest request)
```
Should be `protected virtual`

martijn00 commented on MvvmCross/Windows/Uwp/Views/MvxWindowsViewPresenter.cs:
```diff
@@ -109,49 +173,86 @@ protected virtual string GetRequestText(MvxViewModelRequest request)
         }
 
-        public override void Close(IMvxViewModel viewModel)
+        private bool ClosePage(IMvxViewModel viewModel)
```
Should be `protected virtual`

martijn00 commented on MvvmCross/Windows/Uwp/Views/MvxWindowsViewPresenter.cs:
```diff
@@ -109,49 +173,86 @@ protected virtual string GetRequestText(MvxViewModelRequest request)
+        }
 
-            MvxTrace.Warning("Hint ignored {0}", hint.GetType().Name);
+        private bool CloseSplitView(IMvxViewModel viewModel)
```
Should be `protected virtual`

martijn00 commented on TestProjects/Playground/Playground.Forms.Uwp/Setup.cs:
```diff
@@ -0,0 +1,41 @@
+
+        protected override IMvxWindowsViewPresenter CreateViewPresenter(IMvxWindowsFrame rootFrame)
+        {
+            var presenter = new MvxFormsUwpViewPresenter(rootFrame, FormsApplication);
```
This should not be overriden here.

martijn00 commented on TestProjects/Playground/Playground.Uwp/project.json:
```diff
@@ -1,9 +1,10 @@
   },
   "frameworks": {
-    "uap10.0.10586": {}
+    "uap10.0.10240": {}
```
Please consolidate all UWP projects to "uap10.0.10586"

nickrandolph replied:
@martijn00 is there a reason we want 10586, rather than 10240?

martijn00 replied:
After this we want to update to fall creators update to support .NET Standard 2.0. Other then that we have all Uwp projects target uap10.0.10586 but set minimum to uap10.0.10240

nickrandolph replied:
@martijn00 did you mean for target to be 10586 or the creators update.
Will set min to 10240

nickrandolph commented:
@Cheesebaron @martijn00  I've merged the second presenter and added MvxRegionPresentationAttribute. Have also adjusted UWP min and target version. Let me know what other changes are required for this PR
